pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -475,14 +482,35 @@ func (nc NodeController) tryUpdateNodeStatus(node *api.Node) (error, time.Durati
 			LastTransitionTime: node.CreationTimestamp,
 		}
 		gracePeriod = nodeStartupGracePeriod
+		nc.nodeStatusMap[node.Name] = NodeStatusData{
+			status:    node.Status,
+			timestamp: node.CreationTimestamp,
+		}
 	} else {
 		// If ready condition is not nil, make a copy of it, since we may modify it in place later.
 		lastReadyCondition = *readyCondition
 		gracePeriod = nodeMonitorGracePeriod
 	}
 
-	// Check last time when NodeReady was updated.
-	if nc.now().After(lastReadyCondition.LastProbeTime.Add(gracePeriod)) {
+	oldNodeStatus, found := nc.nodeStatusMap[node.Name]
+	if !found {
+		glog.Errorf("NodeController state corrupted: timestamp map misses entry for %s. Assuming now.", node.Name)
+		oldNodeStatus = NodeStatusData{
+			status:    node.Status,
+			timestamp: nc.now(),
+		}
+		nc.nodeStatusMap[node.Name] = oldNodeStatus
+	} else if !api.Semantic.DeepEqual(oldNodeStatus.status, node.Status) {
+		// If Status changes since the last time we checked, we update the timestamp to "now".
+		glog.V(3).Infof("Status has changed. Updating timestamp: %+v\n vs %+v.", oldNodeStatus.status, node.Status)
+		nc.nodeStatusMap[node.Name] = NodeStatusData{
+			status:    node.Status,
+			timestamp: nc.now(),
+		}
+	}
+
+	// Check last time when NodeController noticed NodeStatus update.
+	if nc.now().After(oldNodeStatus.timestamp.Add(gracePeriod)) {
```

Comment:
Why do you use oldNodeStatus.timestamp instead of current value of nc.nodeStatusMap[node.Name] ?